### PR TITLE
Exchange - Add Max Spend for Fill Order 

### PR DESCRIPTION
**Pull request by gcbsumid**
In order to account for possible slippage in orders, we include a "Max Spend" in the FillBuyOrder or FillSellOrder functions.

If an FillBuy/SellOrder transaction has orders that have been filled before the transaction is able to finish, it may end up slipping and increasing the amount of tokens spent. 

Order 1 - $10 for 3 assets
Order 2 - $20 for 3 assets

Fill Sell Order - 4 items

Ideally, it fills in 3 items from order 1 and 1 item from order 2 and spend $40. But if Order 1 gets filled before this transaction is executed, it may end up buying more items from Order 2 and spend $60 for 3 items instead.

#38

### CODE REVIEW

File: contracts/exchange/Orderbook.sol [hunk 1]
```diff
@@ -139,6 +144,40 @@ contract Orderbook is IOrderbook, ManagerBase {
...
         }
+
+        // get amounts ordered based on AmountToFill and Max Spend
+        amountFilled = 0;
+        uint256 amountSpentOnOrder = 0;
+        for (uint256 i = 0; i < orderAmounts.length; ++i) {
+            if (orderAmounts[i] > 0) {
+                amountSpentOnOrder = orders[_orderIds[i]].price * orderAmounts[i];
+                
+                // Check if there's you're still under tha max Spend
```
Comment (livingaftermidnight) [inline contracts/exchange/Orderbook.sol:155]: Comment has typo

File: contracts/exchange/Orderbook.sol [hunk 2]
```diff
@@ -139,6 +144,40 @@ contract Orderbook is IOrderbook, ManagerBase {
...
+                if (maxSpend == 0) {
+                    orderAmounts[i] = 0;
+                    continue;
+                } else if (maxSpend > amountSpentOnOrder) {
+                    maxSpend -= amountSpentOnOrder;
+                } else if (maxSpend < amountSpentOnOrder) {
+                    orderAmounts[i] = maxSpend / orders[_orderIds[i]].price;
```
Comment (livingaftermidnight) [inline contracts/exchange/Orderbook.sol:162]: Is it possible for price to be zero? Maybe we need a require price >= 0 clause in here.

Comment (livingaftermidnight) [inline contracts/exchange/Orderbook.sol:162]: Is it okay for orderAmounts[i] to store a non-whole number?

File: contracts/exchange/Orderbook.sol [hunk 3]
```diff
@@ -139,6 +144,40 @@ contract Orderbook is IOrderbook, ManagerBase {
...
+                    maxSpend = 0;
+                }
+
+                if (orderAmounts[i] <= amountToFill) {
```
Comment (livingaftermidnight) [inline contracts/exchange/Orderbook.sol:166]: What if orderAmounts[i] == 0 and amountToFill == 0? You would get into this first if statement. If that's not desired, I'd add a "&& amountToFill > 0) check to the if.

File: contracts/exchange/Orderbook.sol (hunk 2 above)
Comment (livingaftermidnight) [reply contracts/exchange/Orderbook.sol:162]: Oh I see it's uint256 so now let me rephrase the question, is it okay for it to truncate the decimal in the event there is one?

File: test/exchange/ExchangeTests.js [hunk 4]
```diff
@@ -132,7 +132,7 @@ describe('Exchange Contract', () => {
 
     it('Supports the Exchange Interface', async () => {
         // IExchange Interface
-        expect(await exchange.supportsInterface("0xdf858c9f")).to.equal(true);
+        expect(await exchange.supportsInterface("0x581b76ff")).to.equal(true);
```
Comment (livingaftermidnight) [inline test/exchange/ExchangeTests.js:135]: For my knowledge, how are you getting this value? Is this just a hash of all the function signatures into one string?

File: test/exchange/OrderbookTests.js [hunk 5]
```diff
@@ -61,7 +61,7 @@ describe('Orderbook Contract tests', () => {
 
     it('Supports the Address resolver Interface', async () => {
       // IOrderbook Interface
-      expect(await orderbook.supportsInterface("0x0950d870")).to.equal(true);
+      expect(await orderbook.supportsInterface("0x2a24311c")).to.equal(true);
```
Comment (livingaftermidnight) [inline test/exchange/OrderbookTests.js:64]: Is this hash value the same as the "orderbook_hash" variable?

**Review by livingaftermidnight — CHANGES_REQUESTED** [review]
Putting this back in Request Changes so we can discuss my comments first before check-in.

File: contracts/exchange/Orderbook.sol (hunk 3 above)
Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:166]: both being equal to zero doesn't really matter here since it'll subtract 0 from amountToFill and add 0 to amountFilled. 

But I get what you're saying, it'll be a waste of gas. I'll do a 'break' to exit immediately if amountToFill is already at 0 or maxSpend is already at 0.

Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:166]: Ignore my last statement. 

This is fine. if orderamounts[i] == 0 and amountToFill == 0 will only occur if maxSpend == 0 (which exists early) or maxSpend is less than amountSpendOnOrder. If it's the second case, orderAmount[i] is set to 0 and it'll only hit this ONCE. This is expected and a non-issue. There's no getting around it. it'll subtract 0 from amountToFill and add 0 to amountFilled so it'll be a no-op anyway.

**Review by butterfiolee — CHANGES_REQUESTED** [review]

File: contracts/exchange/Orderbook.sol [hunk 6]
```diff
@@ -139,6 +144,40 @@ contract Orderbook is IOrderbook, ManagerBase {
...
+
+        // get amounts ordered based on AmountToFill and Max Spend
+        amountFilled = 0;
+        uint256 amountSpentOnOrder = 0;
+        for (uint256 i = 0; i < orderAmounts.length; ++i) {
+            if (orderAmounts[i] > 0) {
+                amountSpentOnOrder = orders[_orderIds[i]].price * orderAmounts[i];
+                
+                // Check if there's you're still under tha max Spend
+                if (maxSpend == 0) {
```
[the remark this answers is not in the document]
Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:156]: oops, you're right. In line 159, it should be maxSpend >= amountSpentOrder. Good catch, that one almost slipped

File: contracts/exchange/Orderbook.sol (hunk 2 above)
Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:162]: i don't know why, but it seems my comments are gone somehow. Reposting comment: 

Price cannot be set to zero. In PlaceOrder(), there's a check to make sure amounts and price is never set to zero. So if the order exists, the price cannot be zero.

Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:162]: We want to round down here. 

I.e. if max spend is 100, and price is 30, we want the orderAmount for this index to round down to 3 items.

Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:162]: Note that, this assumes that the orders are in price ascending / descending order (depending on buy/sell). So in the example above, 30 is the lowest possible price and that's why we're setting maxSpend = 0. The next order _should_ be equal or higher than price 30.

File: test/exchange/ExchangeTests.js (hunk 4 above)
Comment (gcbsumid) [reply test/exchange/ExchangeTests.js:135]: I used console.log() to print out the "type(IExchange).interfaceId" and then copied it over to this test. 

you're correct, it's the hash of all function signatures of the contract

File: test/exchange/OrderbookTests.js (hunk 5 above)
Comment (gcbsumid) [reply test/exchange/OrderbookTests.js:64]: No, this interface id is the hash for all function signatures in the IOrderbook interface. This changes if we update the contract interface.

The CONTRACT_HASH has to be static and non-changing because it's the hash id of the contract in the AddressRegistry (used by the frontend and internal contracts). We can't use the interface id here because it changes over time.

**Review by livingaftermidnight — APPROVED** [review]

**Review by butterfiolee — APPROVED** [review]
